### PR TITLE
Convert all ParcelableReferencable implementations to protobufs

### PR DESCRIPTION
**Pull request by csilvestrini**
I couldn't easily break this down into smaller steps, pretty much had to convert all Referencable subtypes in one go.

First PR in the journey to convert all of these. Hopefully this speeds things up too.

### CODE REVIEW

**Comment by csilvestrini** [comment]
cl/305404330

File: java/arcs/android/crdt/CrdtEntityReferenceProto.kt [hunk 1]
```diff
@@ -0,0 +1,15 @@
+package arcs.android.crdt
+
+import android.os.Parcel
+import arcs.android.util.readProto
+import arcs.core.crdt.CrdtEntity
+
+/** Constructs a [CrdtEntity.ReferenceImpl] from the given [CrdtEntityReferenceProto]. */
+fun fromProto(proto: CrdtEntityReferenceProto) = CrdtEntity.ReferenceImpl(proto.id)
```
Comment (jasonwyatt) [inline java/arcs/android/crdt/CrdtEntityReferenceProto.kt:8]: Can this be an extension function on CrdtEntityReferenceProto?

E.g. `fun CrdtEntityReferenceProto.toReferenceImpl()`

Comment (jasonwyatt) [reply java/arcs/android/crdt/CrdtEntityReferenceProto.kt:8]: Then the call-site below could drop the `let`.

```kotlin
readProto(...)?.toReferenceImpl()
```

File: java/arcs/android/crdt/RawEntityProto.kt [hunk 2]
```diff
@@ -0,0 +1,45 @@
+package arcs.android.crdt
+
+import android.os.Parcel
+import arcs.android.util.readProto
+import arcs.core.data.RawEntity
+
+/** Constructs a [RawEntity] from the given [RawEntityProto]. */
+fun fromProto(proto: RawEntityProto): RawEntity {
```
Comment (jasonwyatt) [inline java/arcs/android/crdt/RawEntityProto.kt:8]: Also - would be nice for this to be an Extension Function on `RawEntityProto`.

File: java/arcs/android/crdt/ReferencablePrimitiveProto.kt [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package arcs.android.crdt
+
+import android.os.Parcel
+import arcs.android.util.readProto
+import arcs.core.data.util.ReferencablePrimitive
+
+/** Constructs a [ReferencablePrimitive] from the given [ReferencablePrimitiveProto]. */
+fun fromProto(proto: ReferencablePrimitiveProto): ReferencablePrimitive<*> = requireNotNull(
```
Comment (jasonwyatt) [inline java/arcs/android/crdt/ReferencablePrimitiveProto.kt:8]: Another spot for an extension fun

File: java/arcs/android/crdt/ReferencableProto.kt [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+import android.os.Parcel
+import arcs.android.util.readProto
+import arcs.core.common.Referencable
+import arcs.core.crdt.CrdtEntity
+import arcs.core.data.RawEntity
+import arcs.core.storage.Reference
+
+/** Constructs a [Referencable] from the given [ReferencableProto]. */
+fun fromProto(proto: ReferencableProto): Referencable? = when (proto.referencableCase) {
```
Comment (jasonwyatt) [inline java/arcs/android/crdt/ReferencableProto.kt:11]: And here.. noticing a pattern - I'll just stop here

File: java/arcs/android/crdt/ReferencableProto.kt [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+/** Serializes a [Referencable] to its proto form. */
+fun Referencable.toProto(): ReferencableProto {
+    val proto = ReferencableProto.newBuilder()
+    when (this) {
+        is RawEntity -> proto.rawEntity = toProto()
+        is CrdtEntity.ReferenceImpl -> proto.crdtEntityReference = toProto()
+        is Reference -> proto.reference = toProto()
+        else -> throw UnsupportedOperationException("Unsupported Referencable: $this.")
```
Comment (jasonwyatt) [inline java/arcs/android/crdt/ReferencableProto.kt:29]: I think we'll need ReferencablePrimitive here too

Comment (csilvestrini) [reply java/arcs/android/crdt/ReferencableProto.kt:29]: Done

File: java/arcs/android/crdt/CrdtEntityReferenceProto.kt (hunk 1 above)
Comment (csilvestrini) [reply java/arcs/android/crdt/CrdtEntityReferenceProto.kt:8]: Good idea, done

File: java/arcs/android/crdt/RawEntityProto.kt (hunk 2 above)
Comment (csilvestrini) [reply java/arcs/android/crdt/RawEntityProto.kt:8]: Done

File: java/arcs/android/crdt/ReferencableProto.kt (hunk 5 above)
Comment (csilvestrini) [reply java/arcs/android/crdt/ReferencableProto.kt:29]: Good catch, done

File: java/arcs/android/crdt/ReferencablePrimitiveProto.kt (hunk 3 above)
Comment (csilvestrini) [reply java/arcs/android/crdt/ReferencablePrimitiveProto.kt]: Done

File: java/arcs/android/crdt/ReferencableProto.kt (hunk 4 above)
Comment (csilvestrini) [reply java/arcs/android/crdt/ReferencableProto.kt]: Done

**Pull request merged**